### PR TITLE
TS-5107: Changes to autolinking in Docutils 1.3

### PR DESCRIPTION
**Pull request by jablko**
Restore the autolinking code and update it for Docutils 1.3, which added the init_customizations() method. It works with Docutils 1.2 and 1.3 now.

@jbfavre already fixed this in #1296, thanks! I realize it would've been much better if I'd had this ready at the beginning of the week ...

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1309/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1203/ for details.

**Comment by zwoop** [comment]
We need into 7.1.x as well, right?

**Comment by jablko** [comment]
@zwoop Yes, I'd like to backport this to 7.1.x if @SolidWallOfCode, @jsime, and @jbfavre approve it. #1296 is in 7.1.x already.

**Comment by jbfavre** [comment]
Well, at least I won't go against it :)

When fixing it in first place, docutils devs feedback was that we wrongly used Inliner, in the sense it's supposed to be internal.
What if another docutils internal change breaks it again ? Don't think it won't happen because the discussion I had showed that not every docutils dev is happy with last change. So, chances are that there'll be another refacto, maybe for 0.15 or 0.16.
Better solution, in both docutils devs & myself, should be to make Sphinx allow customization.

Second concern is that, IMHO, we shouldn't allow 2 different syntax to reference trafficserver issues in documentation.
This usually leads to confusion (why should I use one more than another ?)

Finally, the fix I made has another positive effect: we now have a consistent way of referencing issues, for both Jira & Github. For now, I don't see any way to differentiate Jira & github ones with the Autolink feature.

My 2 cents,

**Comment by jablko** [comment]
> The ``inliner`` parameter takes an instance of `states.Inliner` or a subclass.
> It handles inline markup recognition. A common extension is the addition of
> further implicit hyperlinks, like "RFC 2822". This can be done by subclassing
> `states.Inliner`, adding a new method for the implicit markup, and adding a
> ``(pattern, method)`` pair to the "implicit_dispatch" attribute of the
> subclass. See `states.Inliner.implicit_inline()` for details.

https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/parsers/rst/__init__.py#l59

This is exactly what we're doing. I think it's okay that Docutils added the init_customizations() method between versions -- it sounds like it was their only option.

I don't have a strong sense yet how we want to reference GitHub issues (are there any cases yet?) so it's hard to judge whether consistency between Jira and GitHub is desirable. One thing I personally like about TS-1234 (for Jira) is that we use the same in commit messages, source code comments, etc. so it's consistent in that sense.

**Comment by jbfavre** [comment]
AFAIK, there're no case yet where an issue exists only on Github.
But, this has been "announced" by Bryan Call on trafficserver-dev: https://mail-archives.apache.org/mod_mbox/trafficserver-dev/201612.mbox/%3cE5D8E071-8DE0-4930-BC2D-15CAE1D57777@apache.org%3e so I asume this will come to reality one day.

Like I said, I won't go against the Autolink feature :)

File: doc/conf.py [hunk 1]
```diff
@@ -166,9 +166,71 @@
...
+
+BaseInliner = states.Inliner
+class Inliner(BaseInliner):
+  if hasattr(states.Inliner, 'init_customizations'):
+    def init_customizations(self, settings):
+      self.__class__ = BaseInliner
+      BaseInliner.init_customizations(self, settings)
+      self.__class__ = Inliner
+
+      # Copied from states.Inliner.init_customizations().
```
Comment (SolidWallOfCode) [inline doc/conf.py]: Why is this copied? Isn't it handled by the just previous call to `BaseInliner.init_customization`?

Comment (jablko) [inline doc/conf.py]: In Docutils 0.13 start_string_prefix and end_string_suffix are locals. I'll improve this comment.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1216/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1322/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1239/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1346/ for details.

**Comment by zwoop** [comment]
Status on this? is thing going to land for 7.1.x ?

**Comment by jsime** [comment]
The changes are fine and everything builds fine - I don't see any adverse effects from the inliner for Jira issues. I'm okay with merging this in for 7.1. I'm not sure we'd necessarily end up with comparable inline markup recognition for GitHub issues down the road (`TS-\d+` is pretty distinctive, but `#\d+` or other similar variants that come to mind briefly don't have that same quality).

If there are no objections from anyone else, I'll merge this later this morning.

**Comment by zwoop** [comment]
This doesn't merge cleanly into 7.1.x branch (conflicts on the python script). @jablko Can you make a new PR, for the 7.1.x branch please?

**Comment by jablko** [comment]
@zwoop Sure thing.

**Pull request merged**